osdep/mac/log_helper.swift

```diff
@@ -50,12 +51,23 @@ class LogHelper {
     }
 
     func send(message: String, type: Int) {
-        guard let log = log, UnsafeRawPointer(log).load(as: UInt8.self) != 0 else {
-            logger.log(level: loggerMapping[type] ?? .default, "\(message, privacy: .public)")
-            return
+        lock.withLock {
```

Comment:
This will lock on every message, even if it wouldn't be printed. Is there not a place that we can free log safely after mpv_handle is freed?

Reply:
tough one, theoretical the logging can be called from a few different threads at any time, even at shutdown. atm we also don't have any function that is called when mpv shuts down, besides the shutdown event used here. the closest we have is `func initInput(_ input: OpaquePointer?)`, that is called with nil on shutdown.

the threads that are manly involved in this are the main thread/queue (event handling, usually all cocoa system events) and the playback thread (functions called from core).

though admittedly not everything in the apphub atm is completely thread safe.

Reply:
What if we put `ta_free(UnsafeMutablePointer(log))` in `AppHub::deinit`? This should be after anything else?

Reply:
deinit won't be called on app shut down (memory cleanup) sadly.

https://forums.swift.org/t/deinit-is-not-called-from-main-code/49088
https://forums.swift.org/t/retain-count-set-to-2-and-no-deinit-called/43571
there are many more threads and discussions about this.

Reply:
If it is expected and explicitly done that swift is not freeing resources on program exit. We should do the same, no? At least for logger there is nothing important to do on deinit other than freeing some memory that would be freed anyway on exit.

I think putting "free" in deinit would be ok and if it doesn't trigger it is expected that it was not suppose to be freed and program exits anyway.

I'm fine with either way, but I don't see an issue with not freeing the logger on program exit. That said it may or may not upset leak checkers.

Reply:
that was basically my suggestion in my initial comment.

> alternatively but unclean, we could not set the mpv_handle as ta_parent and not free it at all. it's still possible for uses after free but it won't happen the way the log is used.